### PR TITLE
JBIDE-20842: Initial npm implementation

### PR DESCRIPTION
**Pull request by ibuziuk**
"npm Init" New wizard for package.json creation
"npm Update" & "npm Install" launch shortcuts
Initial SWTBot tests (org.jboss.tools.jst.js.node.test plugin)

### CODE REVIEW

File: plugins/org.jboss.tools.jst.js.npm/.gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+/bin
+/target/
```
Comment (alexeykazakov) [inline plugins/org.jboss.tools.jst.js.npm/.gitignore]: No need to add bin and target to the .gitignore. They both are ignored globally for the jst repo (and other jbt repos too btw).

File: plugins/org.jboss.tools.jst.js.npm/.settings/org.eclipse.jdt.core.prefs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+eclipse.preferences.version=1
```
Comment (alexeykazakov) [inline plugins/org.jboss.tools.jst.js.npm/.settings/org.eclipse.jdt.core.prefs]: I would prefer to not include .settings to the individual projects. Let's use the workspace settings instead. We might want to use a project specific settings in some cases. But this is not the case.

File: plugins/org.jboss.tools.jst.js.npm/src/org/jboss/tools/jst/js/npm/internal/launch/shortcut/GenericNpmLaunch.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+				} else {
+					// Trying to find package.json file ignoring "node_modules" (default modules dir that can not be changed)
+					workingDir = NpmUtil.getBowerWorkingDir(project, NpmConstants.NODE_MODULES);
+				}
+			}
+		}
+		return workingDir;
+	}
+	
+	private void launchBower(IResource resource) throws CoreException {
```
Comment (alexeykazakov) [inline plugins/org.jboss.tools.jst.js.npm/src/org/jboss/tools/jst/js/npm/internal/launch/shortcut/GenericNpmLaunch.java]: should be launchNpm I guess.

File: plugins/org.jboss.tools.jst.js.npm/src/org/jboss/tools/jst/js/npm/internal/preference/NpmPreferencePage.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+	}
+
+	@Override
+	public void init(IWorkbench workbench) {
+		setPreferenceStore(NpmPreferenceHolder.getStore());
+	}
+
+	@Override
+	protected void createFieldEditors() {
+		NpmHomeFieldEditor bowerHomeEditor = new NpmHomeFieldEditor(NpmPreferenceHolder.PREF_NPM_LOCATION,
```
Comment (alexeykazakov) [inline plugins/org.jboss.tools.jst.js.npm/src/org/jboss/tools/jst/js/npm/internal/preference/NpmPreferencePage.java]: bowerHode...->nmpHome...

File: plugins/org.jboss.tools.jst.js.npm/src/org/jboss/tools/jst/js/npm/internal/wizard/NpmInitPage.java [hunk 5]
```diff
@@ -0,0 +1,536 @@
...
+
+	private void entriesChanged() {
+		String dir = dirText.getText();
+		if (dir == null || dir.isEmpty()) {
+			setPageComplete(false);
+			setErrorMessage(Messages.NpmInitWizard_ErrorDirNotDefiened);
+		} else {
+			IContainer container = ResourcesPlugin.getWorkspace().getRoot().getContainerForLocation(new Path(dir));
+			if (container != null && container.exists()) {
+				IFile bowerJson = container.getFile(new Path(NpmConstants.PACKAGE_JSON));
```
Comment (alexeykazakov) [inline plugins/org.jboss.tools.jst.js.npm/src/org/jboss/tools/jst/js/npm/internal/wizard/NpmInitPage.java]: boweJson->packageJson

File: plugins/org.jboss.tools.jst.js.npm/src/org/jboss/tools/jst/js/npm/util/NpmUtil.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+		return gson.toJson(packageJson);
+	}
+	
+	
+	/**
+	 * @return absolute path to directory in which native npm call must be performed. Basically, the method scans 
+	 * project for package.json file and returns it's parent, ignoring "node_modules"
+	 * @throws CoreException
+	 */
+	public static String getBowerWorkingDir(IProject project, final String... ignores) throws CoreException {
```
Comment (alexeykazakov) [inline plugins/org.jboss.tools.jst.js.npm/src/org/jboss/tools/jst/js/npm/util/NpmUtil.java]: getNpmWorkingDir

**Comment by alexeykazakov** [comment]
It looks like the UI is blocked while the launch is being performed?

**Comment by ibuziuk** [comment]
@alexeykazakov thanks for review - fixed!
In terms of blocked UI - I have never seen this behaviour on any OS. If you are keep facing this issue please create a jira and I will try to tackle it

**Pull request merged**